examples/pna-example-tcp-connection-state-tracking-simplified.p4

```diff
@@ -0,0 +1,249 @@
+// in the TCP connection table (ct_tcp_table).
+// Although the same naming is purposedly used for the connection
+// table, the removal of entries (due to time out or explicit connection
+// tear-down is explicitly not supported for the sake of keeping the code
+// simpler and more legible.
+// A complete implementation of TCP connection and status tracking
+// would require to include such functionalities that can be found in
+//  pna-example-tcp-connection-tracking.p4
+
+
+// NOTE: the code is currently written for a compiler supporting
+// if statements within actions. The same functionality could be rewritten
+// for a compiler not supporting if statements within actions by using if
+// if statements in the control block and a table to set up proper context.
+
+
+header ethernet_t {
+    bit<48> dstAddr;
+    bit<48> srcAddr;
+    bit<16> etherType;
+}
+
+header ipv4_t {
+    bit<4>  version;
+    bit<4>  ihl;
+    bit<8>  diffserv;
+    bit<16> totalLength;
+    bit<16> identification;
+    bit<3>  flags;
+    bit<13> fragOffset;
+    bit<8>  ttl;
+    bit<8>  protocol;
+    bit<16> hdrChecksum;
+    bit<32> srcAddr;
+    bit<32> dstAddr;
+}
+
+header tcp_t {
+    bit<16> srcPort;
+    bit<16> dstPort;
+    bit<32> seqNo;
+    bit<32> ackNo;
+    bit<4>  dataOffset;
+    bit<4>  res;
+    bit<8>  flags;
+    bit<16> window;
+    bit<16> checksum;
+    bit<16> urgentPtr;
+}
+
+// Masks of the bit positions of some bit flags within the TCP flags
+// field.
+const bit<8> TCP_ACK_MASK = 0x10;
+const bit<8> TCP_SYN_MASK = 0x02;
+
+// Define names for different expire time profile id values.
+
+const ExpireTimeProfileId_t EXPIRE_TIME_PROFILE_TCP_NOW    = (ExpireTimeProfileId_t) 0;
+const ExpireTimeProfileId_t EXPIRE_TIME_PROFILE_TCP_NEW    = (ExpireTimeProfileId_t) 1;
+const ExpireTimeProfileId_t EXPIRE_TIME_PROFILE_TCP_ESTABLISHED = (ExpireTimeProfileId_t) 2;
+const ExpireTimeProfileId_t EXPIRE_TIME_PROFILE_TCP_NEVER  = (ExpireTimeProfileId_t) 3;
+
+//////////////////////////////////////////////////////////////////////
+// Struct types for holding user-defined collections of headers and
+// metadata in the P4 developer's program.
+//////////////////////////////////////////////////////////////////////
+
+struct metadata_t {
+}
+
+struct headers_t {
+    ethernet_t eth;
+    ipv4_t     ipv4;
+    tcp_t      tcp;
+}
+
+parser MainParserImpl(
+    packet_in pkt,
+    out   headers_t  hdr,
+    inout metadata_t meta,
+    in    pna_main_parser_input_metadata_t istd)
+{
+    state start {
+        pkt.extract(hdr.eth);
+        transition select (hdr.eth.etherType) {
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        pkt.extract(hdr.ipv4);
+        transition select (hdr.ipv4.protocol) {
+            6: parse_tcp;
+            default: accept;
+        }
+    }
+    state parse_tcp {
+        pkt.extract(hdr.tcp);
+        transition accept;
+    }
+}
+
+struct ct_tcp_table_hit_params_t {
```

Comment:
if this struct is used as action parameter, can you show how to specify inout property on individual fields?
Will it be on entire struct? We need ability to specify this per member of the struct.

Reply:
As discussed during the meeting, I think per-action-parameter `rmw` direction is much more easily achieved than doing it on a per-struct-member basis.

Also, this particular struct is very specific to the definition of `add_entry`.  Maybe there is a reasonable way in P4 to define `add_entry` so that it can take a variable number of parameters, varying with the action that one is adding, but the creation of a struct like this for each such `add_entry` call on a different action name is unfortunately part of the current proposal for `add_entry`.  A bit klunky, I know, and if you have better suggestions that fit within the existing P4 language spec, or minor extensions to it, let me know.